
<!DOCTYPE html>
<html>
<head>
  
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">

    <title>Command line reference &mdash; My Docs</title>

    <link rel="shortcut icon" href="../img/favicon.ico">
    <link rel="stylesheet" href="../css/alabaster.css" type="text/css">
    <link rel="stylesheet" href="../css/alabaster-overrides.css" type="text/css">

    

    
      <script src="../search/main.js"></script>
    

    

    <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

    
  
</head>
<body>

  <div class="document">
    <div class="documentwrapper">
      <div class="bodywrapper">
        <div class="body" role="main">
          
            <h2 id="command-line-reference">Command-line reference</h2>
<ul>
<li><strong>All the example dataset and result files are able to find at <a href="http://mbdata.science.ru.nl/qxu/ananse/ananse.html"><strong><em>http://mbdata.science.ru.nl/qxu/ananse/ananse.html</em></strong></a>.</strong></li>
</ul>
<h3 id="build-tf-binding-network">Build TF binding network</h3>
<ul>
<li>
<p>Example:</p>
<p><code>$ ananse binding  -r data/krt_enhancer.bed \
                  -o results/binding.txt \
                  -a /data/hg38_genes.bed \
                  -g hg38 \
                  -p /data/gimme.vertebrate.v5.1.pfm</code></p>
</li>
<li>
<p>Required arguments:</p>
<ul>
<li><code>-r, --enhancers</code><br />
  The name of the input enhancer peak file. This should be a BED format file, with 4 columns. The first column is chromosome name, the second and third columns are the start and end point of peak. We recommend all peaks have 200bp. If the peak is not 200bp, we will normize it to 200bp. The fourth column is intensity of the peak, it could be RPKM or equivalent value. <a href="https://github.com/vanheeringen-lab/ANANSE/raw/master/test/data/krt_enhancer.bed"><strong><em>This</em></strong></a> is an example enhancer BED file.</li>
<li><code>-o, --output</code><br />
  The name of the output file.</li>
</ul>
</li>
<li>
<p>Optional arguments:</p>
<ul>
<li><code>-n, --ncore</code><br />
  Specifies the number of threads to use during analysis.  </li>
<li><code>-g, --genome</code><br />
  The genome that is used for the gene annotation and the enhancer location. 
  This can be either the name of a genome installed with <a href="https://github.com/vanheeringen-lab/genomepy">genomepy</a>, for example <code>hg38</code>, or the name of a genome FASTA file, for example <code>/data/genomes/hg38/hg38.fa</code>. It is recommended to use a genome installed by <code>genomepy</code>. The default setting is <code>hg19</code>.</li>
<li><code>-a, --annotation</code><br />
  The input 12 columns BED file with gene annotation in your genome version. <a href="https://github.com/vanheeringen-lab/ANANSE/raw/master/data/hg38_genes.bed"><strong><em>This</em></strong></a> is an example BED annotation file of human hg38.</li>
<li><code>-p, --motifs</code><br />
  The input Motif file. <a href="/data/gimme.vertebrate.v5.1.pfm"><strong><em>This</em></strong></a> is an example Motif file in vertebrate. if provided there should also be a motif2factors.txt file and a factortable.txt file in the same folder. <a href="/data/gimme.vertebrate.v5.1.motif2factors.txt"><strong><em>This</em></strong></a> is an example of motif2factors file. <a href="/data/gimme.vertebrate.v5.1.factortable.txt"><strong><em>This</em></strong></a> is an example of factortable file.</li>
<li><code>-f, --filter_promoter</code><br />
  Filter promoters, True or False, input should be
  either 'True' or 'False'. (Default setting: True; if 'True', the function will filtered all promoter peaks (+-2k from TSS) in provided enhancer peaks.).</li>
<li><code>-d, --keep_detail</code><br />
  Keep detail files, True or False, input should be either 'True' or 'False'. (Default setting: True).  </li>
<li><code>-h, --help</code><br />
  Show the help message and exit.</li>
</ul>
</li>
</ul>
<h3 id="built-gene-regulatory-network">Built gene regulatory network</h3>
<ul>
<li>
<p>Example:</p>
<p><code>$ ananse network  -e data/KRT_rep1_TPM.txt data/KRT_rep2_TPM.txt \
                  -r data/krt_enhancer.bed \
                  -b results/binding.txt \
                  -o results/full_features.txt \
                  -a /data/hg38_genes.bed \
                  -g hg38 \
                  -c expressioncorrelation.txt \
                  -p ../data/gimme.vertebrate.v5.1.pfm</code></p>
</li>
<li>
<p>Required arguments:</p>
<ul>
<li>
<p><code>-e, --expression</code><br />
  The expression file of your interested cell type or tissue. It could have one or more gene expression file(s). In this file, the 1st column should contain gene name, and a column should be named TPM. <a href="/test/data/KRT_rep1_TPM.txt"><strong><em>This</em></strong></a> is an example of expression file.   </p>
</li>
<li>
<p><code>-b, --binding</code><br />
  The binding network from <code>Build binding network</code> step. One of the example <code>binding network</code> could be found at <a href="http://mbdata.science.ru.nl/qxu/ananse/results/binding.txt"><strong><em>here</em></strong></a>.  </p>
</li>
<li><code>-o, --output</code><br />
  The folder to save results, <code>-o</code> is the required arguments. </li>
</ul>
</li>
<li>
<p>Optional arguments:</p>
<ul>
<li><code>-n, --ncore</code><br />
  Specifies the number of threads to use during analysis. </li>
<li><code>-g, --genome</code><br />
  The genome of your data. For example, hg38. The genome is recommended to download by <code>genomepy</code>.</li>
<li><code>-p, --motifs</code><br />
  The input Motif file. <a href="/data/gimme.vertebrate.v5.1.pfm"><strong><em>This</em></strong></a> is an example Motif file in vertebrate. if provided there should also be a motif2factors.txt file and a factortable.txt file in the same folder. <a href="/data/gimme.vertebrate.v5.1.motif2factors.txt"><strong><em>This</em></strong></a> is an example of motif2factors file. <a href="/data/gimme.vertebrate.v5.1.factortable.txt"><strong><em>This</em></strong></a> is an example of factortable file. </li>
<li><code>-a, --annotation</code><br />
  The input 12 columns BED file with gene annotation in your genome version. <a href="/data/hg38_genes.bed"><strong><em>This</em></strong></a> is an example BED annotation file of human hg38.</li>
<li><code>-f, --filter_promoter</code><br />
  Filter promoters, True or False, input should be
  either 'True' or 'False'. (Default setting: True; if 'True', the function will filtered all promoter peaks (+-2k from TSS) in provided enhancer peaks.).</li>
<li><code>-c, --corrfiles</code><br />
  All gene correlation file, the human gene expression correlation can be found at <a href="http://mbdata.science.ru.nl/qxu/ananse/data/expressioncorrelation.txt"><strong><em>here</em></strong></a>.</li>
<li><code>-h, --help</code><br />
  Show the help message and exit.</li>
</ul>
</li>
</ul>
<h3 id="infer-tf-influence-score">Infer TF influence score</h3>
<ul>
<li>
<p>Example:  </p>
<p><code>$ ananse influence  -a results/full_network.txt \
                    -e data/FB_rep1_TPM.txt \
                    -d data/FB2KRT_degenes.csv \
                    -o results/FB2KRT.txt \
                    -p False</code></p>
</li>
<li>
<p>Required arguments:</p>
<ul>
<li><code>-a, --anetwork</code><br />
The network in second cell. It is the result from <code>Built GRN</code> step. One of the example <code>network</code> could be found at <a href="http://mbdata.science.ru.nl/qxu/ananse/results/full_network.txt"><strong><em>here</em></strong></a>.   </li>
<li><code>-d, --degenes</code><br />
The differential expression table between two cells. <a href="/test/data/FB2KRT_degenes.csv"><strong><em>This</em></strong></a> is an example of differential expression file.  </li>
<li><code>-o, --output</code><br />
The folder to save results, <code>-o</code> is the required arguments.   </li>
</ul>
</li>
<li>
<p>Optional arguments:</p>
<ul>
<li><code>-n, --ncore</code><br />
  Specifies the number of threads to use during analysis. </li>
<li><code>-s, --edges</code><br />
  Specifics the number of top edges (interactions) used. </li>
<li><code>-b, --bnetwork</code><br />
The network in first cell (optional). It is the result from <code>Built GRN</code> step. One of the example <code>network</code> could be found at <a href="http://mbdata.science.ru.nl/qxu/ananse/results/full_network.txt"><strong><em>here</em></strong></a>.  </li>
<li><code>-e, --expression</code><br />
The gene expression in first cell (optional). One or more gene expression file(s), 1st column should contain gene name, and a column should be named TPM. <a href="/test/data/FB_rep1_TPM.txt"><strong><em>This</em></strong></a> is an example of expression file. </li>
<li><code>-p, --plot</code><br />
Plot influence. True or False, input should be either 'True' or 'False'. (Default setting: True)  </li>
<li><code>-h, --help</code><br />
Show the help message and exit.</li>
</ul>
</li>
</ul>
            
          
        </div>
      </div>
    </div>
    <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
      <div class="sphinxsidebarwrapper">
        
          
            
  <h1 class="logo"><a href="..">My Docs</a></h1>



          
            



<h3>Table Of Contents</h3>

<nav>
  
  
    <ul>
    
      <li><a href="#build-tf-binding-network">Build TF binding network</a></li>
      <ul>
    
  </ul>
    
      <li><a href="#built-gene-regulatory-network">Built gene regulatory network</a></li>
      <ul>
    
  </ul>
    
      <li><a href="#infer-tf-influence-score">Infer TF influence score</a></li>
      <ul>
    
  </ul>
    
  </ul>
  

  
</nav>
          
            
  <h3>Related Topics</h3>
  <ul>
    
      <li>Previous: <a href="../API_documentation/" title="previous chapter">
        API documentation
      </a></li>
    
    
      <li>Next: <a href="../examples/" title="next chapter">
        Examples
      </a></li>
    
  </ul>

          
            <div id="searchbox" style="display: none;" role="search">
  <h3>Quick search</h3>
  <form class="search" action="../search.html" method="get">
    <input name="q" type="text">
    <input value="Go" type="submit">
  </form>
  <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
  </p>
</div>
<script type="text/javascript">
  document.getElementById("searchbox").style.display = "block";
</script>
          
        
      </div>
    </div>
    <div class="clearer"></div>
  </div>

  
    <div class="footer">
      
      
        
        Powered by <a href="http://www.mkdocs.org">mkdocs 1.1</a>
        &amp; <a href="https://github.com/iamale/mkdocs-alabaster">mkdocs-alabaster</a>
      
    </div>
  

  <!--
  MkDocs version      : 1.1
  Docs Build Date UTC : 2020-04-10 09:23:33
  -->
</body>
</html>