

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>&lt;no title&gt; &mdash; ANANSE v0.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> ANANSE
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul class="simple">
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ANANSE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>&lt;no title&gt;</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/API_documentation.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <p>## API document</p>
<p>### Working with <cite>Binding</cite> Class
The <cite>Binding</cite> class include all functions used to infer TF-binding site for all enhancer peaks.</p>
<p><a href="#id1"><span class="problematic" id="id2">``</span></a><a href="#id3"><span class="problematic" id="id4">`</span></a>python
import pandas as pd
import dask.dataframe as dd</p>
<p>from ananse import binding
<a href="#id5"><span class="problematic" id="id6">``</span></a><a href="#id7"><span class="problematic" id="id8">`</span></a></p>
<p><code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">gene_bed</span> <span class="pre">=</span> <span class="pre">&quot;/data/hg38_genes.bed&quot;</span>
<span class="pre">peak_bed</span> <span class="pre">=</span> <span class="pre">&quot;data/krt_enhancer.bed&quot;</span>
<span class="pre">pfmfile</span> <span class="pre">=</span> <span class="pre">&quot;/data/gimme.vertebrate.v5.1.pfm&quot;</span>
<span class="pre">`</span></code></p>
<p>A new <cite>Binding</cite> object <cite>bid</cite>
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">bid=binding.Binding(genome=&quot;hg38&quot;,</span> <span class="pre">gene_bed=</span> <span class="pre">gene_bed,</span> <span class="pre">pfmfile=pfmfile)</span>
<span class="pre">`</span></code></p>
<p>Using <cite>clear_peak()</cite> function, we can filter the peaks in promoter ranges.
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">filter_bed</span> <span class="pre">=</span> <span class="pre">bid.clear_peak(peak_bed)</span>
<span class="pre">`</span></code></p>
<p>The <cite>get_PWMScore()</cite> function, calculate the motif z-score of all TFs in all peaks.
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">pfm_weight</span> <span class="pre">=</span> <span class="pre">bid.get_PWMScore(filter_bed)</span>
<span class="pre">pfm</span> <span class="pre">=</span> <span class="pre">dd.read_csv(pfm_weight,</span> <span class="pre">sep=&quot;\t&quot;)</span>
<span class="pre">`</span></code></p>
<p>Load enhancer peak intensity.
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">peak_weight</span> <span class="pre">=</span> <span class="pre">bid.get_peakRPKM(filter_bed)</span>
<span class="pre">peak</span> <span class="pre">=</span> <span class="pre">dd.read_csv(peak_weight,</span> <span class="pre">sep=&quot;\t&quot;)</span>
<span class="pre">`</span></code></p>
<p>Infer the TF-binding score.
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">table=bid.get_binding_score(pfm,</span> <span class="pre">peak)</span>
<span class="pre">`</span></code></p>
<p>The <cite>peak_bed</cite> file is the enhancer peak file, and <cite>“./”</cite> is the output dir. With <cite>run_binding()</cite> function, we can infer the binding sites of all TFs in <cite>peak_bed</cite> enhancer ranges.</p>
<p><code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">bid.run_binding(peak_bed,&quot;./&quot;)</span>
<span class="pre">`</span></code></p>
<p>### Working with <cite>Network</cite> Class
The <cite>Network</cite> class include all functions used to infer cell type specific gene regulatory network.</p>
<p><a href="#id9"><span class="problematic" id="id10">``</span></a><a href="#id11"><span class="problematic" id="id12">`</span></a>python
import pandas as pd
import dask.dataframe as dd</p>
<p>from ananse import network
<a href="#id13"><span class="problematic" id="id14">``</span></a><a href="#id15"><span class="problematic" id="id16">`</span></a></p>
<p><code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">fin_expression</span> <span class="pre">=</span> <span class="pre">[&quot;data/KRT_rep1_TPM.txt&quot;,</span> <span class="pre">&quot;data/KRT_rep2_TPM.txt&quot;]</span>
<span class="pre">binding</span> <span class="pre">=</span> <span class="pre">&quot;results/binding.txt&quot;</span>
<span class="pre">gene_bed</span> <span class="pre">=</span> <span class="pre">&quot;/data/hg38_genes.bed&quot;</span>
<span class="pre">peak_bed</span> <span class="pre">=</span> <span class="pre">&quot;data/krt_enhancer.bed&quot;</span>
<span class="pre">pfmfile</span> <span class="pre">=</span> <span class="pre">&quot;/data/gimme.vertebrate.v5.1.pfm&quot;</span>
<span class="pre">corrfiles</span> <span class="pre">=</span> <span class="pre">&quot;expressioncorrelation.txt&quot;</span>
<span class="pre">`</span></code></p>
<p>A new <cite>Network</cite> object <cite>net</cite>
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">net=network.Network(genome=&quot;hg38&quot;,</span> <span class="pre">gene_bed=</span> <span class="pre">gene_bed,</span> <span class="pre">pfmfile=pfmfile)</span>
<span class="pre">`</span></code></p>
<p>Using <cite>clear_peak()</cite> function, we can filter the peaks in promoter ranges.
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">filter_bed</span> <span class="pre">=</span> <span class="pre">net.clear_peak(peak_bed)</span>
<span class="pre">`</span></code></p>
<p>Using <cite>get_promoter_dataframe()</cite> function to overlap the enhancer with promoter, and using <cite>get_gene_dataframe()</cite> function to overlap the enhancer with gene.
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">prom</span> <span class="pre">=</span> <span class="pre">net.get_promoter_dataframe(filter_bed)</span>
<span class="pre">p</span> <span class="pre">=</span> <span class="pre">net.get_gene_dataframe(filter_bed)</span>
<span class="pre">`</span></code></p>
<p>Using <cite>weight()</cite> function to establish weight function 100 kb up and downstream to the TSS.
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">weight</span> <span class="pre">=</span> <span class="pre">net.distance_weight()</span>
<span class="pre">`</span></code></p>
<p>Using <cite>aggregate_binding()</cite> function to merge binding data with gene.
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">features</span> <span class="pre">=</span> <span class="pre">net.aggregate_binding(binding,</span> <span class="pre">prom,</span> <span class="pre">p,</span> <span class="pre">weight)</span>
<span class="pre">`</span></code></p>
<p>Using <cite>aggregate_binding()</cite> function to establish and merge expression data with features, and using <cite>get_correlation()</cite> function to establsih and merge expression data with features.
<a href="#id17"><span class="problematic" id="id18">``</span></a><a href="#id19"><span class="problematic" id="id20">`</span></a>python
expression_file = net.get_expression(fin_expression, features)
corr_file = net.get_correlation(corrfiles, features)</p>
<dl class="simple">
<dt>other = [</dt><dd><p>expression_file,
corr_file,</p>
</dd>
</dl>
<p>]</p>
<p>featurefile = net.join_features(features, other)
<a href="#id21"><span class="problematic" id="id22">``</span></a><a href="#id23"><span class="problematic" id="id24">`</span></a></p>
<p>Using <cite>create_network()</cite> function to establish gene regulatory network.
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">net.create_network(featurefile,</span> <span class="pre">outfile)</span>
<span class="pre">`</span></code></p>
<p>### Working with <cite>influence</cite> Class</p>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Quan Xu

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>