## Input data

### Genome
When you need run `ANANSE` in your sample, genome file is necessary. We recommand you download your genome file with [genomepy](https://github.com/vanheeringen-lab/genomepy), which is a python package could install genome easily.

In ANANSE, for each genome, we need:

* A genome fasta file
* A 12 columns BED file with genome annotation. 

When you would like to insall the genome with `genomepy`, you could fellow this commands, which will download both `fasta` file and `bed` file.

``` bash
conda create -n genomepy python=3
conda activate genomepy
conda install genomepy=0.7.2

conda activate genompy

# install hg38 genome
genomepy install GRCh38 NCBI -a
```

### Motif database
By default ANANSE uses a non-redundant, clustered database of known vertebrate motifs: `gimme.vertebrate.v5.0`. These motifs come from CIS-BP (http://cisbp.ccbr.utoronto.ca/) and other sources. Large-scale benchmarks using ChIP-seq peaks show that this database shows good performance and should be a good default choice. 

!!! warning
    If you would like to use your own motif database, please makesure your database include following two files: 1, **Motif file**; 2, **Motif2factors file**.
    The `motif` file should end with `.pfm`,  and `motif2factors` file should have the same name with `motif` file and end with `motif2factors.txt`.

* Motif file
```    
#GM.5.0.Sox.0001	M1911_1.02;;M3916_1.02;;M5846_1.02;;M6471_1.02;;M6478_1.02;;MA0084.1_SRY;;MA1152.1_SOX15;;SOX12;;SOX12_1 Sox12_bulyk_sc09-primary;;SOX15_1 Sox15_bulyk_sc09-primary;;SOX18_1 Sox18_bulyk_sc09-primary;;SOX30;;SOX30_1 Sox30_bulyk_sc09-primary;;SOX7_1 Sox7_bulyk_sc09-primary;;SOX9_1 SOX9_transfac_M00410;;SOX9_3 SOX9_jolma_DBD_M136;;SRY;;SRY_3 SRY_jaspar_MA0084.1
>GM.5.0.Sox.0001
0.7213	0.0793	0.1103	0.0891
0.9259	0.0072	0.0062	0.0607
0.0048	0.9203	0.0077	0.0672
0.9859	0.0030	0.0030	0.0081
0.9778	0.0043	0.0128	0.0051
0.1484	0.0050	0.0168	0.8299
#GM.5.0.Homeodomain.0001	M4064_1.02;;TGIF1;;TGIF1_1 TGIF_transfac_M00418;;TGIF1_HUMAN.H11MO.0.A
>GM.5.0.Homeodomain.0001
0.8870	0.0000	0.0178	0.0951
0.1156	0.2033	0.6629	0.0181
0.0017	0.7452	0.0809	0.1722
0.0011	0.0003	0.0003	0.9983
0.0026	0.0141	0.9721	0.0111
0.0000	0.0189	0.0054	0.9758
0.0006	0.9983	0.0006	0.0006
0.9170	0.0140	0.0046	0.0644
0.2228	0.2421	0.3300	0.2051
0.3621	0.1054	0.2208	0.3116
0.5727	0.0104	0.1741	0.2428
```

* Motif2factors file  
```
Motif	Factor	Evidence	Curated
GM.5.0.Sox.0001	SRY	JASPAR	Y
GM.5.0.Sox.0001	SOX9	Transfac	Y
GM.5.0.Sox.0001	Sox9	Transfac	N
GM.5.0.Sox.0001	SOX9	SELEX	Y
GM.5.0.Sox.0001	Sox9	SELEX	N
GM.5.0.Sox.0001	SOX13	ChIP-seq	Y
GM.5.0.Sox.0001	SOX9	ChIP-seq	Y
GM.5.0.Sox.0001	Sox9	ChIP-seq	N
GM.5.0.Sox.0001	SRY	SELEX	Y
```

!!! note "Example"  
    The default motif database (`gimme.vertebrate.v5.0`) can be found at [GimmeMotif](https://github.com/vanheeringen-lab/gimmemotifs) package[^1] at here:  

    * [gimme.vertebrate.v5.0.pfm](https://github.com/vanheeringen-lab/gimmemotifs/blob/master/data/motif_databases/gimme.vertebrate.v5.0.pfm)  
    * [gimme.vertebrate.v5.0.motif2factors.txt](https://github.com/vanheeringen-lab/gimmemotifs/blob/master/data/motif_databases/gimme.vertebrate.v5.0.motif2factors.txt)

### Enhancer data
The EP300 ChIP-seq or ATAC-seq peaks were used to define putative enhancer regions for specific cell types. The summits of the MACS2 peaks were chosen first, then extended (+/- 100bp) to a total size of 200bp. The EP300 or H3K27ac ChIP-seq peak intensity generated by MACS2 (bedGraph file) was used to represent enhancer intensity. bedGraphToBigWig tool[^2] was used to switch bedGraph files to bigWig files. The intensity of enhancer peak was calculated by bigWigSummary tool[^2], which selected the highest signal of EP300 (200 bp around the peak summit) or H3K27ac (2,000 bp around the peak summit) ChIP-seq peak.

This is the example of input BED file:

```
chr2	148881617	148881817	7
chr7	145997204	145997404	4
chr7	145997304	145997404	4
chr7	145997104	145997404	4
chr13	109424160	109424360	20
chr14	32484901	32485101	2
```

!!! note "Example"
    You can find our test samples enhancer files here: 

    * [FB_enhancer.bed](https://github.com/vanheeringen-lab/ANANSE/blob/master/test/data/FB_enhancer.bed)  
    * [KRT_enhancer.bed](https://github.com/vanheeringen-lab/ANANSE/blob/master/test/data/KRT_enhancer.bed)

### Expression data
The expression data normally from RNA-seq experiment. We are using `TPM` score to represent gene's expression in ANANSE. In expression file, the 1st column (named as `target_id`) should contain **gene name**, and the second column should be named as `tpm`. 

This is the example of input expression file:

```
target_id	tpm
A1BG	9.579771
A1CF	0.0223
A2M	0.25458
A2ML1	664.452
A3GALT2	0.147194
```

!!! note "Example"
    You can find our test samples enhancer files here:  

    * [FB_rep1_TPM.txt](/test/data/FB_rep1_TPM.txt)  
    * [FB_rep2_TPM.txt](/test/data/FB_rep2_TPM.txt)  
    * [KRT_rep1_TPM.txt](/test/data/KRT_rep1_TPM.txt)  
    * [KRT_rep2_TPM.txt](/test/data/KRT_rep2_TPM.txt)  

### Differential expression data
The differential expression data normally from RNA-seq experiment. In differential expression file, the 1st column (named as `resid`) should contain **gene name**, the second column should be named as `log2FoldChange` which is **log2 FoldChange score** of gene, and the third column should be named as `padj`, which is **p-value** of gene. 

This is the example of input expression file:

```
resid	log2FoldChange	padj
ANPEP	7.44242618323665	0
CD24	-8.44520139575174	0
COL1A2	8.265875689393	0
COL6A1	7.41947749996406	0
COL6A2	9.62485194293185	0
COL6A3	11.0553152937569	0
DAB2	7.46610079411987	0
DMKN	-11.6435948368453	0
```
!!! warning
    The `log2FoldChange` should **negative number** if this gene is up regulated, and **positive number** if this gene is down regulated.

!!! note "Example"
    You can find our test samples enhancer files here:  

    * [FB2KRT_degenes.csv](/test/data/FB2KRT_degenes.csv)  

[^1]: van Heeringen, S.J., and Veenstra, G.J.C. (2010). GimmeMotifs: a de novo motif prediction pipeline for ChIP-sequencing experiments. Bioinformatics 27, 270-271.
[^2]: Kempfer, R., and Pombo, A. (2019). Methods for mapping 3D chromosome architecture. Nat Rev Genet.
Kent, J., ENCODE DCC. (2014). kentUtils: Jim Kent command line bioinformatic utilities. Available from: https://github.com/ENCODE-DCC/kentUtils.
