
<!DOCTYPE html>
<html>
<head>
  
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">

    <title>API documentation &mdash; My Docs</title>

    <link rel="shortcut icon" href="../img/favicon.ico">
    <link rel="stylesheet" href="../css/alabaster.css" type="text/css">
    <link rel="stylesheet" href="../css/alabaster-overrides.css" type="text/css">

    

    
      <script src="../search/main.js"></script>
    

    

    <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

    
  
</head>
<body>

  <div class="document">
    <div class="documentwrapper">
      <div class="bodywrapper">
        <div class="body" role="main">
          
            <h2 id="api-document">API document</h2>
<h3 id="working-with-binding-class">Working with <code>Binding</code> Class</h3>
<p>The <code>Binding</code> class include all functions used to infer TF-binding site for all enhancer peaks.</p>
<pre><code class="python">import pandas as pd
import dask.dataframe as dd

from ananse import binding
</code></pre>

<pre><code class="python">gene_bed = &quot;/data/hg38_genes.bed&quot;
peak_bed = &quot;data/krt_enhancer.bed&quot;
pfmfile = &quot;/data/gimme.vertebrate.v5.1.pfm&quot;
</code></pre>

<p>A new <code>Binding</code> object <code>bid</code></p>
<pre><code class="python">bid=binding.Binding(genome=&quot;hg38&quot;, gene_bed= gene_bed, pfmfile=pfmfile)
</code></pre>

<p>Using <code>clear_peak()</code> function, we can filter the peaks in promoter ranges. </p>
<pre><code class="python">filter_bed = bid.clear_peak(peak_bed)
</code></pre>

<p>The <code>get_PWMScore()</code> function, calculate the motif z-score of all TFs in all peaks.</p>
<pre><code class="python">pfm_weight = bid.get_PWMScore(filter_bed)
pfm = dd.read_csv(pfm_weight, sep=&quot;\t&quot;)
</code></pre>

<p>Load enhancer peak intensity.</p>
<pre><code class="python">peak_weight = bid.get_peakRPKM(filter_bed)
peak = dd.read_csv(peak_weight, sep=&quot;\t&quot;)
</code></pre>

<p>Infer the TF-binding score.</p>
<pre><code class="python">table=bid.get_binding_score(pfm, peak)
</code></pre>

<p>The <code>peak_bed</code> file is the enhancer peak file, and <code>"./"</code> is the output dir. With <code>run_binding()</code> function, we can infer the binding sites of all TFs in <code>peak_bed</code> enhancer ranges.</p>
<pre><code class="python">bid.run_binding(peak_bed,&quot;./&quot;)
</code></pre>

<h3 id="working-with-network-class">Working with <code>Network</code> Class</h3>
<p>The <code>Network</code> class include all functions used to infer cell type specific gene regulatory network.</p>
<pre><code class="python">import pandas as pd
import dask.dataframe as dd

from ananse import network
</code></pre>

<pre><code class="python">fin_expression = [&quot;data/KRT_rep1_TPM.txt&quot;, &quot;data/KRT_rep2_TPM.txt&quot;]
binding = &quot;results/binding.txt&quot;
gene_bed = &quot;/data/hg38_genes.bed&quot;
peak_bed = &quot;data/krt_enhancer.bed&quot;
pfmfile = &quot;/data/gimme.vertebrate.v5.1.pfm&quot;
corrfiles = &quot;expressioncorrelation.txt&quot;
</code></pre>

<p>A new <code>Network</code> object <code>net</code></p>
<pre><code class="python">net=network.Network(genome=&quot;hg38&quot;, gene_bed= gene_bed, pfmfile=pfmfile)
</code></pre>

<p>Using <code>clear_peak()</code> function, we can filter the peaks in promoter ranges. </p>
<pre><code class="python">filter_bed = net.clear_peak(peak_bed)
</code></pre>

<p>Using <code>get_promoter_dataframe()</code> function to overlap the enhancer with promoter, and using <code>get_gene_dataframe()</code> function to overlap the enhancer with gene.</p>
<pre><code class="python">prom = net.get_promoter_dataframe(filter_bed)
p = net.get_gene_dataframe(filter_bed)
</code></pre>

<p>Using <code>weight()</code> function to establish weight function 100 kb up and downstream to the TSS.</p>
<pre><code class="python">weight = net.distance_weight()
</code></pre>

<p>Using <code>aggregate_binding()</code> function to merge binding data with gene.</p>
<pre><code class="python">features = net.aggregate_binding(binding, prom, p, weight)
</code></pre>

<p>Using <code>aggregate_binding()</code> function to establish and merge expression data with features, and using <code>get_correlation()</code> function to establsih and merge expression data with features.</p>
<pre><code class="python">expression_file = net.get_expression(fin_expression, features)
corr_file = net.get_correlation(corrfiles, features)

other = [
    expression_file,
    corr_file,
]

featurefile = net.join_features(features, other)
</code></pre>

<p>Using <code>create_network()</code> function to establish gene regulatory network.</p>
<pre><code class="python">net.create_network(featurefile, outfile)
</code></pre>

<h3 id="working-with-influence-class">Working with <code>influence</code> Class</h3>
            
          
        </div>
      </div>
    </div>
    <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
      <div class="sphinxsidebarwrapper">
        
          
            
  <h1 class="logo"><a href="..">My Docs</a></h1>



          
            



<h3>Table Of Contents</h3>

<nav>
  
  
    <ul>
    
      <li><a href="#working-with-binding-class">Working with Binding Class</a></li>
      <ul>
    
  </ul>
    
      <li><a href="#working-with-network-class">Working with Network Class</a></li>
      <ul>
    
  </ul>
    
      <li><a href="#working-with-influence-class">Working with influence Class</a></li>
      <ul>
    
  </ul>
    
  </ul>
  

  
</nav>
          
            
  <h3>Related Topics</h3>
  <ul>
    
      <li>Previous: <a href=".." title="previous chapter">
        Home
      </a></li>
    
    
      <li>Next: <a href="../command-line_reference/" title="next chapter">
        Command line reference
      </a></li>
    
  </ul>

          
            <div id="searchbox" style="display: none;" role="search">
  <h3>Quick search</h3>
  <form class="search" action="../search.html" method="get">
    <input name="q" type="text">
    <input value="Go" type="submit">
  </form>
  <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
  </p>
</div>
<script type="text/javascript">
  document.getElementById("searchbox").style.display = "block";
</script>
          
        
      </div>
    </div>
    <div class="clearer"></div>
  </div>

  
    <div class="footer">
      
      
        
        Powered by <a href="http://www.mkdocs.org">mkdocs 1.1</a>
        &amp; <a href="https://github.com/iamale/mkdocs-alabaster">mkdocs-alabaster</a>
      
    </div>
  

  <!--
  MkDocs version      : 1.1
  Docs Build Date UTC : 2020-04-10 09:23:33
  -->
</body>
</html>